Redesign the try_reverse_output to support more cases

zhuqi-lucas opened the pull request:
## Which issue does this PR close?
- Closes #19403

## Rationale for this change

After #19064 was completed, I found it couldn't meet our internal project requirements:

1. **Redesign try_reverse_output using EquivalenceProperties**
   - The previous implementation used a simple `is_reverse` method that could only handle basic reverse matching
   - Now leveraging `EquivalenceProperties` can handle more complex scenarios, including constant column elimination and monotonic functions

2. **Switch to ordering_satisfy method for ordering matching**
   - This method internally:
     - Normalizes orderings (removes constant columns)
     - Checks monotonic functions (like `date_trunc`, `CAST`, `CEIL`)
     - Handles prefix matching

3. **Extend sort pushdown support to more operators**
   - Added `try_pushdown_sort` implementation for `ProjectionExec`, `FilterExec`, `CooperativeExec`
   - These operators can now pass sort requests down to their children

## What changes are included in this PR?

### Core Changes:

1. **ParquetSource::try_reverse_output** (datasource-parquet/src/source.rs)
   - Added `eq_properties` parameter
   - Reverses all orderings in equivalence properties
   - Uses `ordering_satisfy` to check if reversed ordering satisfies the request
   - Removed `file_ordering` field and `with_file_ordering_info` method

2. **FileSource trait** (datasource/src/file.rs)
   - Updated `try_reverse_output` signature with `eq_properties` parameter
   - Added detailed documentation explaining parameter usage and examples

3. **FileScanConfig::try_pushdown_sort** (datasource/src/file_scan_config.rs)
   - Simplified logic to directly call `file_source.try_reverse_output`
   - No longer needs to pre-check ordering satisfaction or set file ordering info

4. **New operator support**
   - `FilterExec::try_pushdown_sort` - Pushes sort below filters
   - `ProjectionExec::try_pushdown_sort` - Pushes sort below projections
   - `CooperativeExec::try_pushdown_sort` - Supports sort pushdown in cooperative execution

5. **Removed obsolete methods**
   - Deleted `LexOrdering::is_reverse` - replaced by `ordering_satisfy`

### Feature Enhancements:

**Supported optimization scenarios:**

1. **Constant column elimination** (Test 7)
```sql
   -- File ordering: [timeframe ASC, period_end ASC]
   -- Query: WHERE timeframe = 'quarterly' ORDER BY period_end DESC
   -- Effect: After timeframe becomes constant, reverse scan is enabled
```

2. **Monotonic function support** (Test 8)
```sql
   -- File ordering: [ts ASC]
   -- Query: ORDER BY date_trunc('month', ts) DESC
   -- Effect: date_trunc is monotonic, reverse scan satisfies the request
```

## Are these changes tested?

Yes, comprehensive tests have been added:

- **Test 7 (237 lines)**: Constant column elimination scenarios
  - Single constant column filter
  - Multi-value IN clauses (doesn't trigger optimization)
  - Literal constants in sort expressions
  - Non-leading column filters (edge cases)

- **Test 8 (355 lines)**: Monotonic function scenarios
  - `date_trunc` (date truncation)
  - `CAST` (type conversion)
  - `CEIL` (ceiling)
  - `ABS` (negative case - not monotonic over mixed positive/negative range)

All tests verify:
- Presence of `reverse_row_groups=true` in physical plans
- Correctness of query results

## Are there any user-facing changes?

**API Changes:**
- `FileSource::try_reverse_output` signature changed (added `eq_properties` parameter)
- Removed `FileSource::with_file_ordering_info` method
- Removed `LexOrdering::is_reverse` public method

**User-visible improvements:**
- More queries can leverage reverse row group scanning for optimization
- Especially queries with `WHERE` clauses that make certain columns constant
- Queries using monotonic functions (like date functions, type conversions)

**Note:** This PR returns `Inexact` results because only row group order is reversed, not row order within row groups. Future enhancements could include:
- File reordering based on statistics (returning `Exact`)
- Partial sort pushdown for prefix matches

zhuqi-lucas commented on datafusion/sqllogictest/test_files/dynamic_filter_pushdown_config.slt:
```diff
@@ -752,3 +752,577 @@ DROP TABLE sorted_parquet;
 SET datafusion.optimizer.enable_sort_pushdown = true;
+
+
+# Test 7: Sort pushdown with constant column filtering
```
Good point @adriangb , I added sort pushdown cases here because it included topk pushdown testing here, i can add a new test file.

adriangb commented:
I'm comfortable with `ProjectionExec` and `CooperativeExec` but I think `FilterExec` needs some more discussion since pushing ordering below filters might have a negative impact on performance for some queries / file sources.

zhuqi-lucas commented:
> I'm comfortable with `ProjectionExec` and `CooperativeExec` but I think `FilterExec` needs some more discussion since pushing ordering below filters might have a negative impact on performance for some queries / file sources.

@adriangb Happy holidays! Thank you for the feedback! I completely understand your concern about FilterExec.

If i makes sense right, i've updated the implementation to address the performance issue you mentioned. Now `FilterExec::try_pushdown_sort` only pushes down the sort when the child node already has an output ordering:

```rust
fn try_pushdown_sort(
        &self,
        order: &[PhysicalSortExpr],
    ) -> Result<SortOrderPushdownResult<Arc<dyn ExecutionPlan>>> {
        if let Some(_child_ordering) = self.input.output_ordering() {
            match self.input.try_pushdown_sort(order)? {
                SortOrderPushdownResult::Exact { inner } => {
                    let new_exec = Arc::new(self.clone()).with_new_children(vec![inner])?;
                    Ok(SortOrderPushdownResult::Exact { inner: new_exec })
                }
                SortOrderPushdownResult::Inexact { inner } => {
                    let new_exec = Arc::new(self.clone()).with_new_children(vec![inner])?;
                    Ok(SortOrderPushdownResult::Inexact { inner: new_exec })
                }
                SortOrderPushdownResult::Unsupported => {
                    Ok(SortOrderPushdownResult::Unsupported)
                }
            }
        } else {
            Ok(SortOrderPushdownResult::Unsupported)
        }
    }
```

This way:
- When the data source can optimize to use pushdown sort, so we pushdown to leverage the existing ordering.
- When the data source is unsorted, we don't pushdown, so sorting happens after filtering on the smaller filtered dataset

This should avoid the performance degradation you were concerned about while still enabling the optimization for sorted data sources. The new tests validate this with the `WHERE timeframe = 'quarterly'` cases, which require pushdown through FilterExec to utilize the sorted Parquet file's ordering.


What do you think about this approach?

adriangb commented:
It seems like that case is just because filter pushdown is not enabled (i.e. #19477). So I think the behavior is correct in this case: we would not want to push the ordering down past the filter.

zhuqi-lucas commented:
> It seems like that case is just because filter pushdown is not enabled (i.e. #19477). So I think the behavior is correct in this case: we would not want to push the ordering down past the filter.

I already enabled it in slt, so it's strange, i will investigate it today.
```
statement ok
SET datafusion.execution.parquet.pushdown_filters = true;
```

adriangb commented:
> > It seems like that case is just because filter pushdown is not enabled (i.e. #19477). So I think the behavior is correct in this case: we would not want to push the ordering down past the filter.
> 
> I already enabled it in slt, so it's strange, i will investigate it today.
> 
> ```
> statement ok
> SET datafusion.execution.parquet.pushdown_filters = true;
> ```

Then it is indeed a mystery to me but I suspect if we can fix that root cause the sort order won't be as much of an issue.

zhuqi-lucas commented:
@adriangb My mistake, i found one line in mid of the file i reset to false about pushdown_filters, fixed it in latest PR.

And also removed sort pushdown logic for filterexec.

adriangb commented on datafusion/physical-plan/src/projection.rs:
```diff
@@ -357,6 +359,27 @@ impl ExecutionPlan for ProjectionExec {
+
+    fn try_pushdown_sort(
+        &self,
+        order: &[PhysicalSortExpr],
```
What if the sort depends on the output of the projection? Presumably then the projection should say the sort can't be pushed down?

adriangb commented on datafusion/physical-plan/src/projection.rs:
```diff
@@ -357,6 +359,69 @@ impl ExecutionPlan for ProjectionExec {
+            let mut can_pushdown = true;
+            let transformed = Arc::clone(&sort_expr.expr).transform(|expr| {
+                if let Some(col) = expr.as_any().downcast_ref::<Column>() {
+                    // Check if column index is valid
```
```suggestion
                    // Check if column index is valid.
                    // This should always be true but fail gracefully if it's not.
```

adriangb commented on datafusion/physical-plan/src/projection.rs:
```diff
@@ -357,6 +359,69 @@ impl ExecutionPlan for ProjectionExec {
+
+                    let proj_expr = &self.expr()[col.index()];
+
+                    // Check if projection expression is a simple column
```
```suggestion
                    // Check if projection expression is a simple column
                    // We cannot push down order by clauses that depend on
                    // projected computations as they would have nothing to reference.
```

adriangb commented on datafusion/core/tests/physical_optimizer/pushdown_sort.rs:
```diff
@@ -670,3 +669,272 @@ fn test_pushdown_through_blocking_node() {
+        Ok:
+          - SortExec: expr=[a@2 DESC NULLS LAST], preserve_partitioning=[false]
+          -   ProjectionExec: expr=[c@2 as c, b@1 as b, a@0 as a]
+          -     DataSourceExec: file_groups={1 group: [[x]]}, projection=[a, b, c, d, e], file_type=parquet, reverse_row_groups=true
```
Nit: for these tests it would be nice to have a TestScan type thing that accepts any order just to see it piped down all the way.

adriangb commented on datafusion/core/tests/physical_optimizer/test_utils.rs:
```diff
@@ -823,3 +825,157 @@ pub fn sort_expr_named(name: &str, index: usize) -> PhysicalSortExpr {
+/// A test data source that can display any requested ordering
+/// This is useful for testing sort pushdown behavior
+#[derive(Debug, Clone)]
+pub struct TestScan {
```
I thought we already had a `TestScan` 🤔

alamb reviewed (commented):
Thanks @zhuqi-lucas @xudong963  and @adriangb  -- I took a quick scan though this PR (I don't have time to review it thoroughly but I took a cursory look over it and it looks good to me

This is so cool

zhuqi-lucas commented on datafusion/datasource/src/file.rs:
```diff
@@ -150,15 +150,46 @@ pub trait FileSource: Send + Sync {
+    ///
+    /// # Parameters
+    /// * `order` - The requested sort ordering
+    /// * `eq_properties` - Equivalence properties that can be used to determine if a reversed
```
Good suggestion @alamb , i added more comments about the eq_properties now in latest commit.

zhuqi-lucas commented:
> Thanks @zhuqi-lucas @xudong963 and @adriangb -- I took a quick scan though this PR (I don't have time to review it thoroughly but I took a cursory look over it and it looks good to me
> 
> This is so cool

Thank you @alamb for review, i also added more comments about the  eq_properties represents in the try_reverse_output function.
